### PR TITLE
[ModuleInliner] use std::make_heap to construct the heap after element updated

### PR DESCRIPTION
**Pull request by taoliq**
After `updateAndCheckDecreased`, some element may be updated and the range(Heap.begin(), Heap.end()) is not a valid heap.
Use std::make_heap to make sure the heap is valid.

### CODE REVIEW

**Comment by kazutakahirata** [comment]
Hmm.  `std::make_heap` in the loop would be expensive.  We would be looking at all elements in the queue every time we pop one item.  Let me take some time and try to understand the underlying issue better.

**Comment by kazutakahirata** [comment]
@taoliq, the following seems to work even with the expensive checks enabled.  I'll post this as separate PR myself:

```
    std::pop_heap(Heap.begin(), Heap.end(), isLess);
    while (updateAndCheckDecreased(Heap.back())) {
      std::push_heap(Heap.begin(), Heap.end(), isLess);
      std::pop_heap(Heap.begin(), Heap.end(), isLess);
    }
    std::push_heap(Heap.begin(), Heap.end(), isLess);
```

Although `pop_heap()` simply swaps the first and last element of the heap, the libstdc++ version (and possibly the C++ standard) requires that the entire range be a valid heap when expensive checks are enabled.

The code snippet above:

- swaps the first and last element while `Heap` is a valid heap,
- calls `updateAndCheckDecreased` on the last element, which is outside the valid heap, and
- put the last element back into the heap. 

In other words, we use the last element as a safe work area while we are updating its priority.

We could reduce the number of calls to `updateAndCheckDecreased` in certain situations, but let's take care of the crash first.

**Comment by kazutakahirata** [comment]
I've posted https://github.com/llvm/llvm-project/pull/69251 as an alternative fix.

**Comment by taoliq** [comment]
> I've posted #69251 as an alternative fix.

Thanks Kazu. I like your fix, so I'd like to close this one.
